File: src/main/java/com/wxl/dyttcrawler/processor/MetaRefreshPageProcessor.java
package com.wxl.dyttcrawler.processor;

import com.wxl.dyttcrawler.urlhandler.PriorityUrlCalculator;
import com.wxl.dyttcrawler.urlhandler.UrlFilter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.nodes.Element;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;
import us.codecraft.webmagic.Page;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Create by wuxingle on 2020/7/18
 * 处理浏览器自动刷新的标签
 * <head>
 * <meta http-equiv="refresh" content="0;URL=index.htm">
 * </head>
 */
@Slf4j
@Order(-100)
@Component
public class MetaRefreshPageProcessor extends AbstractDyttProcessor {

    private static final Pattern LOCATION_PATTERN = Pattern.compile("URL=(.+)");

    @Autowired
    public MetaRefreshPageProcessor(PriorityUrlCalculator priorityCalculator,
                                    UrlFilter urlFilter) {
        super(priorityCalculator, urlFilter);
    }

    @Override
    public void process(Page page) {
        log.debug("process refresh meta:{}", page.getUrl());
        Element refreshMeta = getRefreshMeta(page);
        String content = refreshMeta.attr("content");
        if (StringUtils.isNotBlank(content)) {
            Matcher matcher = LOCATION_PATTERN.matcher(content);
            if (matcher.find()) {
                String location = matcher.group(1);
                addLink(page, location);
            }
        }
    }

    @Override
    public boolean match(Page page) {
        return getRefreshMeta(page) != null;
    }

    private Element getRefreshMeta(Page page) {
        Element head = page.getHtml().getDocument().head();
        return head.select("meta[http-equiv=refresh]").first();
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/processor/MetaRefreshPageProcessor.kt
package com.wxl.dyttcrawler.processor

import com.wxl.dyttcrawler.urlhandler.PriorityUrlCalculator
import com.wxl.dyttcrawler.urlhandler.UrlFilter
import org.jsoup.nodes.Element
import org.slf4j.LoggerFactory
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import us.codecraft.webmagic.Page
import java.util.regex.Pattern

/**
 * Create by wuxingle on 2021/10/12
 * 处理浏览器自动刷新的标签
 * <head>
 * <meta http-equiv="refresh" content="0;URL=index.htm">
 * </head>
 */
@Order(-100)
@Component
class MetaRefreshPageProcessor(
    priorityUrlCalculator: PriorityUrlCalculator,
    urlFilter: UrlFilter
) : AbstractDyttProcessor(priorityUrlCalculator, urlFilter) {

    companion object {
        private val log = LoggerFactory.getLogger(MetaRefreshPageProcessor::class.java)

        private val LOCATION_PATTERN = Pattern.compile("URL=(.+)")
    }

    override fun process(page: Page) {
        log.debug("process refresh meta:{}", page.url)

        val refreshMeta = getRefreshMeta(page)
        val content = refreshMeta?.attr("content")
        if (!content.isNullOrBlank()) {
            val matcher = LOCATION_PATTERN.matcher(content)
            if (matcher.find()) {
                val location = matcher.group(1)
                addLink(page, location)
            }
        }
    }

    override fun match(page: Page): Boolean {
        return getRefreshMeta(page) !== null
    }

    private fun getRefreshMeta(page: Page): Element? {
        val head = page.html.document.head()
        return head.select("meta[http-equiv=refresh]").first()
    }

}